src/java.base/share/classes/java/lang/FdLibm.java

```diff
@@ -504,6 +513,449 @@ static double compute(double y, double x) {
+     *            in (3).
+     *   3. Final rounding
+     *      After generating the 53 bits result, we compute one more bit.
+     *      Together with the remainder, we can decide whether the
+     *      result is exact, bigger than 1/2ulp, or less than 1/2ulp
+     *      (it will never equal to 1/2ulp).
+     *      The rounding mode can be detected by checking whether
+     *      huge + tiny is equal to huge, and whether huge - tiny is
+     *      equal to huge for some floating point number "huge" and "tiny".
+     *
+     * Special cases:
+     *      sqrt(+-0) = +-0         ... exact
+     *      sqrt(inf) = inf
+     *      sqrt(-ve) = NaN         ... with invalid signal
+     *      sqrt(NaN) = NaN         ... with invalid signal for signaling NaN
+     *
+     * Other methods : see the appended file at the end of the program below.
+     *---------------
+     */
+    static class Sqrt {
+        private Sqrt() {throw new UnsupportedOperationException();}
+
+        private static final double tiny = 1.0e-300;
+
+        static double compute(double x) {
+            double z = 0.0;
+            int sign = 0x8000_0000;
+            /*unsigned*/ int r, t1, s1, ix1, q1;
+            int ix0, s0, q, m, t, i;
+
+            ix0 = __HI(x);  // high word of x
+            ix1 = __LO(x);  // low word of x
+
+            // take care of Inf and NaN
+            if((ix0 & 0x7ff0_0000) == 0x7ff0_0000) {
+                return x*x + x; // sqrt(NaN)=NaN, sqrt(+inf)=+inf, sqrt(-inf)=sNaN
+            }
+            // take care of zero
+            if (ix0 <= 0) {
+                if (((ix0 & (~sign)) | ix1) == 0)
+                    return x; // sqrt(+-0) = +-0
+                else if (ix0 < 0)
+                    return (x-x)/(x-x); // sqrt(-ve) = sNaN
+            }
+            // normalize x
+            m = (ix0 >> 20);
+            if (m == 0) { // subnormal x
+                while (ix0 == 0) {
+                    m -= 21;
+                    ix0 |= (ix1 >>> 11); // unsigned shift
+                    ix1 <<= 21;
+                }
+                for(i = 0; (ix0 & 0x0010_0000) == 0; i++) {
+                    ix0 <<= 1;
+                }
+                m -= i-1;
+                ix0 |= (ix1 >>> (32 - i)); // unsigned shift
+                ix1 <<= i;
+            }
+            m -= 1023;      // unbias exponent */
+            ix0 = (ix0 & 0x000f_ffff) | 0x0010_0000;
+            if ((m & 1) != 0){        // odd m, double x to make it even
+                ix0 += ix0 + ((ix1 & sign) >>> 31); // unsigned shift
+                ix1 += ix1;
+            }
+            m >>= 1;        // m = [m/2]
+
+            // generate sqrt(x) bit by bit
+            ix0 += ix0 + ((ix1 & sign) >>> 31); // unsigned shift
+            ix1 += ix1;
+            q = q1 = s0 = s1 = 0;   // [q,q1] = sqrt(x)
+            r = 0x0020_0000;        // r = moving bit from right to left
+
+            while(r != 0) {
```

Comment:
```suggestion
            while (r != 0) {
```

Reply:
Thanks for the careful reading; will fix.